src/backend/optimizer/path/allpaths.c

```diff
@@ -2962,6 +2963,15 @@ set_cte_pathlist(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
 
 	if (!is_shared)
 	{
+		/*
+		 * If plan sharing is disabled, we avoid performing DML inside CTE for
+		 * multi reference case. Otherwise, it'll cause duplicated DML
+		 * operations.
+		 */
+		if (cte->cterefcount > 1 &&
+			cte_query_contains_dml((Node *) cte->ctequery, cteroot))
```

Comment:
Why not using approach from [6X](https://github.com/arenadata/gpdb/pull/553/files#diff-4eff9e11378a8d7655a2cd15eea18c2fa93687163c15bcdd6e4dd54818fef85fR2042)? It covers more cases

Reply:
Yep, these changes shold be backported. Initial patch for 6X didn't take complex cases into account